\chapter{Concorrência em Java}
\label{cha:concjava}

\section{Interface Runnable e classe Thread}

A linguagem de programação \texttt{Java} fornece diretamente suporte a
aplicações \textit{multithread}, permitindo a criação de linhas de
execução por meio de extensão da classe \verb!Thread! ou implementação
da interface \verb!Runnable!.

Não há diferenças, no que diz respeito a concorrência, entre um método
e outro. Porém, como a linguagem disponibiliza suporte apenas a
herança simples, extender a classe \verb!Thread! significa que não
será possível herdar funcionalidades de qualquer outra classe. Para
resolver esse problema, a aplicação pode criar uma hierarquia de
classes ou, então, implementar a interface \verb!Runnable!.

Indiferente do método utilizado, o corpo da classe deve conter a
implementação de um método \verb!run!. Este será invocado ao iniciar a
\textit{thread} por meio do método \verb!start!. O método \verb!run!
serve um propósito similar ao da função \verb!main! na linguagem
\texttt{C}, ou mesmo \texttt{Java}, ou seja, ao encerrar a execução do
mesmo a linha de execução é considerada como encerrada e não pode mais
ser reiniciada.

\section{Synchronized}
% XXX Acho que nem da pra corrigir sem reescrever.

O modificador \textit{synchronized} tem seu uso e importância destacados quando
várias \textit{threads} estão sendo executadas em um programa e estas
necessitam atualizar dados compartilhados. Tais \textit{threads} podem
tentar realizar
essa operação ao mesmo tempo, o que poderia levar a um resultado
inesperado. Para resolver este problema a \textit{Java} disponibiliza o
modificador \textit{synchronized} para implementar a exclusão mutua.

Ao marcar um método com \textit{synchronized}, JVM 
garantirá que apenas uma \textit{thread} acesse tal método. Se outras % xxx acho que isso aqui ta errado
\textit{threads} tentarem fazer o mesmo, elas serão colocadas em espera até que a
\textit{thread} atual finalize seu trabalho e libere o método.

\section{Exception}

A linguagem de programação Java utiliza exceções para tratar erros e
demais eventos excepcionais. Uma exceção é um evento que ocorre durante
a execução de um programa que interrompe o fluxo normal das instruções
\cite{except}.

O primeiro passo na construção de um manipulador de exceção é colocar
o código que pode lançar uma exceção dentro de um bloco \texttt{try}. Em geral,
um bloco \texttt{try} é semelhante ao seguinte.

\begin{lstlisting}

try {
    code
}
catch and finally blocks . . .
    
\end{lstlisting}


Se ocorrer uma exceção no bloco \texttt{try}, a exceção é tratada por um
manipulador de exceção a ele associado. Para associar um manipulador de
exceção com um bloco \texttt{try}, você definir colocar um bloco
\texttt{catch}.

Você associa manipuladores de exceção, com um bloco \texttt{try}, fornecendo
um ou mais blocos \texttt{catch} logo após o bloco \texttt{try}. Nenhum
código pode ser entre o final do bloco \texttt{try} e no início do primeiro bloco % WTF?!
\texttt{catch}.

\begin{lstlisting}

try {
     
} catch (ExceptionType name) {
     
} catch (ExceptionType name) {
     
}  

\end{lstlisting}
